lib/Sema/TypeCheckStmt.cpp

```diff
@@ -749,35 +749,53 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
                                       *conformance,
                                       TC.Context.Id_Iterator,
                                       diag::sequence_protocol_broken);
-
       if (!generatorTy)
         return nullptr;
-      
-      Expr *getIterator
-        = TC.callWitness(sequence, DC, sequenceProto, *conformance,
-                         TC.Context.Id_makeIterator,
-                         {}, diag::sequence_protocol_broken);
-      if (!getIterator) return nullptr;
+
+      auto witness = conformance->getWitnessByName(
+          sequenceType->getRValueType(), TC.Context.Id_makeIterator);
+      if (!witness)
+        return nullptr;
+      S->setMakeGenerator(witness);
 
       // Create a local variable to capture the generator.
       std::string name;
       if (auto np = dyn_cast_or_null<NamedPattern>(S->getPattern()))
         name = "$"+np->getBoundName().str().str();
       name += "$generator";
+
       generator = new (TC.Context)
         VarDecl(/*IsStatic*/false, VarDecl::Specifier::Var, /*IsCaptureList*/false,
                 S->getInLoc(), TC.Context.getIdentifier(name), DC);
       generator->setType(generatorTy);
       generator->setInterfaceType(generatorTy->mapTypeOutOfContext());
       generator->setImplicit();
+      S->setGeneratorVar(generator);
 
-      // Create a pattern binding to initialize the generator.
       auto genPat = new (TC.Context) NamedPattern(generator);
       genPat->setImplicit();
+
+      // TODO: test/DebugInfo/iteration.swift requires this extra info to
+      // be around.
+      auto nextResultType = cast<FuncDecl>(witness.getDecl())
+                                ->getResultInterfaceType()
+                                .subst(witness.getSubstitutions())
+                                ->getCanonicalType();
       auto *genBinding = PatternBindingDecl::createImplicit(
-          TC.Context, StaticSpellingKind::None, genPat, getIterator, DC,
+          TC.Context, StaticSpellingKind::None, genPat,
+          new (TC.Context) OpaqueValueExpr(S->getInLoc(), nextResultType), DC,
           /*VarLoc*/ S->getForLoc());
-      S->setIterator(genBinding);
+
+      // Necessary type coersion for method application.
+      Type newSequenceType = cast<AbstractFunctionDecl>(witness.getDecl())
```

Comment:
Is this not conformance->getType()?

Reply:
for B in:
open class A : Sequence {}
open class B : A {}
I get B using conformace->getType() and A using my method. A is what I want.